OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,286 @@
+/* -------------------------------------------------------------------------- *
+ *                            OpenSim:  DataTable.h                           *
+ * -------------------------------------------------------------------------- *
+ * The OpenSim API is a toolkit for musculoskeletal modeling and simulation.  *
+ * See http://opensim.stanford.edu and the NOTICE file for more information.  *
+ * OpenSim is developed at Stanford University and supported by the US        *
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2015 Stanford University and the Authors                *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** \file
+This file defines the  DataTable_ class, which is used by OpenSim to provide an 
+in-memory container for data access and manipulation.                         */
+
+#ifndef OPENSIM_DATATABLE_H
+#define OPENSIM_DATATABLE_H
+
+// Non-standard headers.
+#include "SimTKcommon.h"
+#include "OpenSim/Common/Exception.h"
+#include "OpenSim/Common/ValueArrayDictionary.h"
+
+namespace OpenSim {
+
+/** AbstractDataTable is the base-class of all DataTable_(templated) allowing 
+storage of DataTable_ templated on different types to be stored in a container 
+like std::vector. AbstractDataTable_ offers:
+- Interface to access column metadata of the DataTable_.
+- A heterogeneous container to store metadata associated with the DataTable_ in
+  the form of key-value pairs where key is of type std::string and value can be
+  of any type.
+
+This class is abstract and cannot be used directly. Create instances of 
+DataTable_ instead. See DataTable_ for details on ussage.                     */
+class AbstractDataTable {
+protected:
+    class ValueDictionary {
```

Comment:
Is this class really worth the additional code? Since it is returned by public methods it would need to be documented. What about just using `ValueArrayDictionary` so there would only be one class and concept to document? Maybe you could add a convenience method or default index value so that element 0 would get returned if no index was specified?